<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8" /><title>Bio.Algorithm.Sequence</title><link href="ocean.css" rel="stylesheet" type="text/css" title="Ocean" /><script src="haddock-util.js" type="text/javascript"></script><script type="text/javascript">//<![CDATA[
window.onload = function () {pageLoad();setSynopsis("mini_Bio-Algorithm-Sequence.html");};
//]]>
</script></head><body><div id="package-header"><ul class="links" id="page-menu"><li><a href="index.html">Contents</a></li><li><a href="doc-index.html">Index</a></li></ul><p class="caption">ba-0.1.0.0: General bioinformatics algorithms library</p></div><div id="content"><div id="module-header"><table class="info"><tr><th>Copyright</th><td>(C) 2014 Ricky Elrod</td></tr><tr><th>License</th><td>BSD2 (see LICENSE file)</td></tr><tr><th>Maintainer</th><td>Ricky Elrod &lt;ricky@elrod.me&gt;</td></tr><tr><th>Stability</th><td>experimental</td></tr><tr><th>Portability</th><td>lens</td></tr><tr><th>Safe Haskell</th><td>Safe-Inferred</td></tr><tr><th>Language</th><td>Haskell2010</td></tr></table><p class="caption">Bio.Algorithm.Sequence</p></div><div id="table-of-contents"><p class="caption">Contents</p><ul><li><a href="#g:1">Complements and Reverse Complements</a></li><li><a href="#g:2">k-mer algorithms</a></li><li><a href="#g:3">Skew</a></li></ul></div><div id="description"><p class="caption">Description</p><div class="doc"><p>Algorithms dealing with bioinformatics sequences. Right now it doesn't do
 much.</p></div></div><div id="synopsis"><p id="control.syn" class="caption expander" onclick="toggleSection('syn')">Synopsis</p><ul id="section.syn" class="hide" onclick="toggleSection('syn')"><li class="src short"><a href="#v:dnaReverseComplement">dnaReverseComplement</a> :: <a href="Bio-Algorithm-Types-Sequence.html#t:DNA">DNA</a> -&gt; <a href="Bio-Algorithm-Types-Sequence.html#t:DNA">DNA</a></li><li class="src short"><a href="#v:rnaReverseComplement">rnaReverseComplement</a> :: <a href="Bio-Algorithm-Types-Sequence.html#t:RNA">RNA</a> -&gt; <a href="Bio-Algorithm-Types-Sequence.html#t:RNA">RNA</a></li><li class="src short"><a href="#v:kmers">kmers</a> :: <a href="Bio-Algorithm-Types-RawSequence.html#t:AsRawSequence">AsRawSequence</a> s =&gt; <a href="file:///usr/share/doc/ghc/html/libraries/base/Data-Int.html#t:Int">Int</a> -&gt; s -&gt; [s]</li><li class="src short"><a href="#v:maxKmers">maxKmers</a> :: (<a href="Bio-Algorithm-Types-RawSequence.html#t:AsRawSequence">AsRawSequence</a> s, <a href="file:///usr/share/doc/ghc/html/libraries/base/Data-Ord.html#t:Ord">Ord</a> s) =&gt; [s] -&gt; [s]</li><li class="src short"><a href="#v:ltClumps">ltClumps</a> :: <a href="file:///usr/share/doc/ghc/html/libraries/base/Data-Int.html#t:Int">Int</a> -&gt; <a href="file:///usr/share/doc/ghc/html/libraries/base/Data-Int.html#t:Int">Int</a> -&gt; <a href="file:///usr/share/doc/ghc/html/libraries/base/Data-Int.html#t:Int">Int</a> -&gt; <a href="Bio-Algorithm-Types-RawSequence.html#t:RawSequence">RawSequence</a> -&gt; [<a href="Bio-Algorithm-Types-RawSequence.html#t:RawSequence">RawSequence</a>]</li><li class="src short"><a href="#v:skews">skews</a> :: <a href="Bio-Algorithm-Types-RawSequence.html#t:AsRawSequence">AsRawSequence</a> s =&gt; s -&gt; [<a href="file:///usr/share/doc/ghc/html/libraries/base/Data-Int.html#t:Int">Int</a>]</li></ul></div><div id="interface"><h1 id="g:1">Complements and Reverse Complements</h1><div class="top"><p class="src"><a name="v:dnaReverseComplement" class="def">dnaReverseComplement</a> :: <a href="Bio-Algorithm-Types-Sequence.html#t:DNA">DNA</a> -&gt; <a href="Bio-Algorithm-Types-Sequence.html#t:DNA">DNA</a></p><div class="doc"><p>Return the reverse complement for some DNA sequence.</p><pre class="screen"><code class="prompt">&gt;&gt;&gt; </code><strong class="userinput"><code>dnaReverseComplement (DNA (RawSequence (BL.pack &quot;CCTTGGAA&quot;)))
</code></strong>DNA (RawSequence &quot;TTCCAAGG&quot;)
</pre><pre class="screen"><code class="prompt">&gt;&gt;&gt; </code><strong class="userinput"><code>T.pack &quot;CCTTGGAA&quot; ^. lazy . _RawSequence . to (dnaReverseComplement . DNA)
</code></strong>DNA (RawSequence &quot;TTCCAAGG&quot;)
</pre></div></div><div class="top"><p class="src"><a name="v:rnaReverseComplement" class="def">rnaReverseComplement</a> :: <a href="Bio-Algorithm-Types-Sequence.html#t:RNA">RNA</a> -&gt; <a href="Bio-Algorithm-Types-Sequence.html#t:RNA">RNA</a></p><div class="doc"><p>Return the reverse complement for some RNA sequence.</p><pre class="screen"><code class="prompt">&gt;&gt;&gt; </code><strong class="userinput"><code>rnaReverseComplement (RNA (RawSequence (BL.pack &quot;GAUGGAACUUGACUACGUAAAUU&quot;)))
</code></strong>RNA (RawSequence &quot;AAUUUACGUAGUCAAGUUCCAUC&quot;)
</pre><pre class="screen"><code class="prompt">&gt;&gt;&gt; </code><strong class="userinput"><code>T.pack &quot;GAUGGAACUUGACUACGUAAAUU&quot; ^. lazy . _RawSequence . to (rnaReverseComplement . RNA)
</code></strong>RNA (RawSequence &quot;AAUUUACGUAGUCAAGUUCCAUC&quot;)
</pre></div></div><h1 id="g:2">k-mer algorithms</h1><div class="top"><p class="src"><a name="v:kmers" class="def">kmers</a></p><div class="subs arguments"><p class="caption">Arguments</p><table><tr><td class="src">:: <a href="Bio-Algorithm-Types-RawSequence.html#t:AsRawSequence">AsRawSequence</a> s</td><td class="doc empty">&nbsp;</td></tr><tr><td class="src">=&gt; <a href="file:///usr/share/doc/ghc/html/libraries/base/Data-Int.html#t:Int">Int</a></td><td class="doc"><p>The length of the k-mers (i.e., the <em>k</em> of the &quot;k-mer&quot;)</p></td></tr><tr><td class="src">-&gt; s</td><td class="doc"><p>The sequence to search in</p></td></tr><tr><td class="src">-&gt; [s]</td><td class="doc"><p>The resulting list of k-mers</p></td></tr></table></div><div class="doc"><p>Find and list all of the kmers of the given length in the given string.</p><pre class="screen"><code class="prompt">&gt;&gt;&gt; </code><strong class="userinput"><code>kmers 3 (RawSequence $ BL.pack &quot;AGATCGAGTG&quot;)
</code></strong>[RawSequence &quot;AGA&quot;,RawSequence &quot;GAT&quot;,RawSequence &quot;ATC&quot;,RawSequence &quot;TCG&quot;,RawSequence &quot;CGA&quot;,RawSequence &quot;GAG&quot;,RawSequence &quot;AGT&quot;,RawSequence &quot;GTG&quot;]
</pre><pre class="screen"><code class="prompt">&gt;&gt;&gt; </code><strong class="userinput"><code>kmers 5 (RawSequence $ BL.pack &quot;GTAGAGCTGT&quot;)
</code></strong>[RawSequence &quot;GTAGA&quot;,RawSequence &quot;TAGAG&quot;,RawSequence &quot;AGAGC&quot;,RawSequence &quot;GAGCT&quot;,RawSequence &quot;AGCTG&quot;,RawSequence &quot;GCTGT&quot;]
</pre><pre class="screen"><code class="prompt">&gt;&gt;&gt; </code><strong class="userinput"><code>&quot;gatatat&quot; ^. _RawSequence . to (kmers 4)
</code></strong>[RawSequence &quot;gata&quot;,RawSequence &quot;atat&quot;,RawSequence &quot;tata&quot;,RawSequence &quot;atat&quot;]
</pre><pre class="screen"><code class="prompt">&gt;&gt;&gt; </code><strong class="userinput"><code>&quot;gatatat&quot; ^. _RawSequence . to DNA . to (kmers 4)
</code></strong>[DNA (RawSequence &quot;gata&quot;),DNA (RawSequence &quot;atat&quot;),DNA (RawSequence &quot;tata&quot;),DNA (RawSequence &quot;atat&quot;)]
</pre><pre class="screen"><code class="prompt">&gt;&gt;&gt; </code><strong class="userinput"><code>&quot;gatatat&quot; ^. _RawSequence . to RNA . to (kmers 4)
</code></strong>[RNA (RawSequence &quot;gata&quot;),RNA (RawSequence &quot;atat&quot;),RNA (RawSequence &quot;tata&quot;),RNA (RawSequence &quot;atat&quot;)]
</pre></div></div><div class="top"><p class="src"><a name="v:maxKmers" class="def">maxKmers</a> :: (<a href="Bio-Algorithm-Types-RawSequence.html#t:AsRawSequence">AsRawSequence</a> s, <a href="file:///usr/share/doc/ghc/html/libraries/base/Data-Ord.html#t:Ord">Ord</a> s) =&gt; [s] -&gt; [s]</p><div class="doc"><p>Return only the kmers that appear the highest number of times.</p><pre class="screen"><code class="prompt">&gt;&gt;&gt; </code><strong class="userinput"><code>&quot;gatatat&quot; ^. _RawSequence . to (maxKmers . kmers 4)
</code></strong>[RawSequence &quot;atat&quot;]
</pre><pre class="screen"><code class="prompt">&gt;&gt;&gt; </code><strong class="userinput"><code>&quot;gatatat&quot; ^. _RawSequence . to (maxKmers . kmers 4 . DNA)
</code></strong>[DNA (RawSequence &quot;atat&quot;)]
</pre><pre class="screen"><code class="prompt">&gt;&gt;&gt; </code><strong class="userinput"><code>&quot;gatatat&quot; ^. _RawSequence . to (maxKmers . kmers 4 . RNA)
</code></strong>[RNA (RawSequence &quot;atat&quot;)]
</pre></div></div><div class="top"><p class="src"><a name="v:ltClumps" class="def">ltClumps</a></p><div class="subs arguments"><p class="caption">Arguments</p><table><tr><td class="src">:: <a href="file:///usr/share/doc/ghc/html/libraries/base/Data-Int.html#t:Int">Int</a></td><td class="doc"><p><code>k</code>. The length of the k-mers (i.e., the <em>k</em> of the &quot;k-mer&quot;)</p></td></tr><tr><td class="src">-&gt; <a href="file:///usr/share/doc/ghc/html/libraries/base/Data-Int.html#t:Int">Int</a></td><td class="doc"><p><code>l</code>. The interval size to scan</p></td></tr><tr><td class="src">-&gt; <a href="file:///usr/share/doc/ghc/html/libraries/base/Data-Int.html#t:Int">Int</a></td><td class="doc"><p><code>t</code>. How many times it must appear</p></td></tr><tr><td class="src">-&gt; <a href="Bio-Algorithm-Types-RawSequence.html#t:RawSequence">RawSequence</a></td><td class="doc"><p>The sequence to search in</p></td></tr><tr><td class="src">-&gt; [<a href="Bio-Algorithm-Types-RawSequence.html#t:RawSequence">RawSequence</a>]</td><td class="doc"><p>The resulting list of k-mers</p></td></tr></table></div><div class="doc"><p>Find and list all distinct (L, t)-clumps of the given length in the given
 string.</p><p>This is done by scanning the entire genome with a window of size <code>l</code>. In each
 window, we get all <code>k</code>-mers, see if they exist <code>t</code> times, and if they do, we
 add them to the result.</p><pre class="screen"><code class="prompt">&gt;&gt;&gt; </code><strong class="userinput"><code>&quot;gatcagcataagggtccctgcaatgcatgacaagcctgcagttgttttac&quot; ^. _RawSequence . to (ltClumps 4 25 3)
</code></strong>[RawSequence &quot;tgca&quot;]
</pre></div></div><h1 id="g:3">Skew</h1><div class="top"><p class="src"><a name="v:skews" class="def">skews</a> :: <a href="Bio-Algorithm-Types-RawSequence.html#t:AsRawSequence">AsRawSequence</a> s =&gt; s -&gt; [<a href="file:///usr/share/doc/ghc/html/libraries/base/Data-Int.html#t:Int">Int</a>]</p><div class="doc"><p>Find skews (difference in number of G and number of C occurances in some
 sequence) by removing a nucleotide from the right side and recursing over the
 remaining sequence.</p><p>Can be used to solve the Rosalind &quot;Minimum Skew Problem&quot; like this:
  <code>minSkewIndices s = map (+1) . findIndices (== minimum (skews s)) $ (skews s)</code>
 although it is too slow to actually be accepted.</p></div></div></div></div><div id="footer"><p>Produced by <a href="http://www.haskell.org/haddock/">Haddock</a> version 2.14.3</p></div></body></html>